gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactor/handler/impl/DefaultReactorHandlerRegistry.java

```diff
@@ -194,85 +177,61 @@ private void removeHttpAcceptors(final Reactable reactable, final List<Reactable
     }
 
     @Override
-    public Collection<HttpAcceptorHandler> getHttpAcceptorHandlers() {
-        return registeredHttpAcceptors;
-    }
+    public <T extends Acceptor<T>> Collection<T> getAcceptors(Class<T> acceptorType) {
+        Collection<T> acceptorsType = (Collection<T>) acceptors.get(acceptorType);
 
-    private void addHttpAcceptors(List<HttpAcceptorHandler> reactableEntrypoints) {
-        synchronized (this) {
-            final List<HttpAcceptorHandler> httpAcceptorHandlers = new ArrayList<>(registeredHttpAcceptors);
-            httpAcceptorHandlers.addAll(reactableEntrypoints);
-            httpAcceptorHandlers.sort(httpAcceptorHandlerComparator);
-
-            registeredHttpAcceptors = httpAcceptorHandlers;
+        if (acceptorsType == null) {
+            return Collections.emptyList();
         }
-    }
-
-    private void remove(List<HttpAcceptorHandler> previousEntrypoints) {
-        synchronized (this) {
-            final List<HttpAcceptorHandler> httpAcceptorHandlers = new ArrayList<>(registeredHttpAcceptors);
-            httpAcceptorHandlers.removeAll(previousEntrypoints);
-            httpAcceptorHandlers.sort(httpAcceptorHandlerComparator);
 
-            registeredHttpAcceptors = httpAcceptorHandlers;
-        }
+        return acceptorsType;
     }
 
-    private static class ReactableHttpAcceptors {
-
-        private final ReactorHandler handler;
+    private void addAcceptors(List<? extends Acceptor> newAcceptors) {
+        if (!newAcceptors.isEmpty()) {
+            synchronized (this) {
+                // When adding new acceptors, we are assuming they are all of the same subtype.
+                Class<? extends Acceptor<?>> acceptorType = (Class<? extends Acceptor<?>>) newAcceptors
+                    .get(0)
+                    .getClass()
+                    .getInterfaces()[0];
```

Comment:
depending on interfaces order could lead to issue. Couldn't you use the concret class instead of one of the interfaces ? or at least get the one assignableTo Acceptor?

Reply:
We can't rely on concrete class, otherwise the dispatcher would have to retrieve the list of acceptors for this concret class, and not the interface.
I was a bit worried to rely on assignableTo from a perf perspective.
But I do agree that it's not optimal.